xls/modules/zstd/buffer.x

```diff
@@ -0,0 +1,451 @@
+
+    let (result3, data3) = buffer_pop_checked(result2.buffer, u32:16);
+    assert_eq(result3, BufferResult {
+        status: BufferStatus::NO_ENOUGH_DATA,
+        buffer: result2.buffer
+    });
+    assert_eq(data3, u32:0);
+}
+
+// Pops `length` amount of data from a `buffer`. Returns the modified buffer
+// and the popped data. The length of the popped data is a function parameter.
+// It will fail if the buffer has no sufficient amount of data.
+// For calls that need better error handling, check `buffer_sized_pop_checked`.
+pub fn buffer_sized_pop<BSIZE: u32, WSIZE: u32> (buffer: Buffer<BSIZE>) -> (Buffer<BSIZE>, bits[WSIZE]) {
+    let (buffer, value) = buffer_pop(buffer, WSIZE);
+    (buffer, value as bits[WSIZE])
+}
+
+#[test]
+fn test_buffer_sized_pop() {
+    let buffer = Buffer { contents: u32:0xDEADBEEF, length: u32:32 };
+    let (buffer, data) = buffer_sized_pop<u32:32, u32:16>(buffer);
+    assert_eq(data, u16:0xBEEF);
+    assert_eq(buffer, Buffer { contents: u32:0xDEAD, length: u32:16 });
+    let (buffer, data) = buffer_sized_pop<u32:32, u32:16>(buffer);
+    assert_eq(data, u16:0xDEAD);
+    assert_eq(buffer, Buffer { contents: u32:0, length: u32:0 });
+}
+
+// Pops `length` amount of data from a `buffer` if possible. The length of
+// the popped data is a function parameter. Returns the status of the operation
+// and the updated buffer in a BufferResult structure, and the popped data.
+pub fn buffer_sized_pop_checked<BSIZE: u32, WSIZE: u32> (buffer: Buffer<BSIZE>) -> (BufferResult<BSIZE>, bits[WSIZE]) {
+    let (result, value) = buffer_pop_checked(buffer, WSIZE);
+    (result, value as bits[WSIZE])
+}
+
+#[test]
+fn test_buffer_sized_pop_checked() {
+    let buffer = Buffer { contents: u32:0xDEADBEEF, length: u32:32 };
+    let (result1, data1) = buffer_sized_pop_checked<u32:32, u32:16>(buffer);
+    assert_eq(result1, BufferResult {
+        status: BufferStatus::OK,
+        buffer: Buffer { contents: u32:0xDEAD, length: u32:16 }
+    });
+    assert_eq(data1, u16:0xBEEF);
+
+    let (result2, data2) = buffer_sized_pop_checked<u32:32, u32:16>(result1.buffer);
+    assert_eq(result2, BufferResult {
+        status: BufferStatus::OK,
+        buffer: Buffer { contents: u32:0, length: u32:0 }
+    });
+    assert_eq(data2, u16:0xDEAD);
+
+    let (result3, data3) = buffer_sized_pop_checked<u32:32, u32:16>(result2.buffer);
+    assert_eq(result3, BufferResult {
+        status: BufferStatus::NO_ENOUGH_DATA,
+        buffer: result2.buffer
+    });
+    assert_eq(data3, u16:0);
+}
+
+// Returns `length` amount of data from a `buffer`.
+// It will fail if the buffer has no sufficient amount of data.
+// For calls that need better error handling, check `buffer_peek_checked`.
+pub fn buffer_peek<BSIZE: u32>(buffer: Buffer<BSIZE>, length: u32) -> bits[BSIZE] {
+    if buffer_has_at_least(buffer, length) == false {
+        trace_fmt!("Not enough data in the buffer!");
+        fail!("not_enough_data", ());
```

Comment:
is there a resonable default we can return in the failing case?

Reply:
I added the same default values as returned by the `_checked` variant of the function (in the failing case).